<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Power User Guide: Efficient Glycan Manipulation</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Power User Guide: Efficient Glycan Manipulation</h1>



<div id="welcome-to-the-advanced-zone" class="section level2">
<h2>Welcome to the Advanced Zone! üöÄ</h2>
<p>Ready to unlock the <strong>full potential</strong> of <code>glyrepr</code>? This vignette is for those who want to peek under the hood and master the art of efficient glycan computation. If you‚Äôre writing custom functions for glycan analysis or building the next great glycomics tool, you‚Äôre in the right place!</p>
<p><strong>Fair warning</strong>: This guide assumes you‚Äôre comfortable with R programming and graph theory concepts. If you‚Äôre just getting started, check out our ‚ÄúGetting Started with glyrepr‚Äù vignette first.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glyrepr)</span></code></pre></div>
</div>
<div id="the-secret-superpower-unique-structure-optimization" class="section level2">
<h2>The Secret Superpower: Unique Structure Optimization</h2>
<p>Before we dive into the <code>smap</code> functions, let‚Äôs understand why they exist and why they‚Äôre <strong>game-changing</strong> for glycan analysis.</p>
<div id="the-problem-glycan-computation-is-expensive" class="section level3">
<h3>The Problem: Glycan Computation is Expensive üí∏</h3>
<p>Working with glycan structures means working with graphs, and graph operations are computationally expensive. When you‚Äôre analyzing thousands of glycans from a large-scale study, this becomes a real bottleneck.</p>
</div>
<div id="the-solution-work-smart-not-hard" class="section level3">
<h3>The Solution: Work Smart, Not Hard üß†</h3>
<p><code>glyrepr</code> implements a clever optimization called <strong>unique structure storage</strong>. Instead of storing thousands of identical graphs, it stores only the unique ones and keeps track of which original positions they belong to.</p>
<p>Let‚Äôs see this in action:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Our test data: some common glycan structures</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>iupacs <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc(b1-&quot;</span>,  <span class="co"># N-glycan core</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;Gal(b1-3)GalNAc(a1-&quot;</span>,                                    <span class="co"># O-glycan core 1</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;Gal(b1-3)[GlcNAc(b1-6)]GalNAc(a1-&quot;</span>,                     <span class="co"># O-glycan core 2</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;Man(a1-3)[Man(a1-6)]Man(a1-3)[Man(a1-6)]Man(a1-&quot;</span>,          <span class="co"># Branched mannose</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;GlcNAc6Ac(b1-4)Glc3Me(a1-&quot;</span>                              <span class="co"># With decorations</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>struc <span class="ot">&lt;-</span> <span class="fu">as_glycan_structure</span>(iupacs)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Now let&#39;s create a realistic dataset with lots of repetition</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>large_struc <span class="ot">&lt;-</span> <span class="fu">rep</span>(struc, <span class="dv">1000</span>)  <span class="co"># 5,000 total structures</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>large_struc</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &lt;glycan_structure[5000]&gt;</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc(b1-</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [2] Gal(b1-3)GalNAc(a1-</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [3] Gal(b1-3)[GlcNAc(b1-6)]GalNAc(a1-</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [4] Man(a1-3)[Man(a1-6)]Man(a1-3)[Man(a1-6)]Man(a1-</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [5] GlcNAc6Ac(b1-4)Glc3Me(a1-</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [6] Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc(b1-</span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [7] Gal(b1-3)GalNAc(a1-</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [8] Gal(b1-3)[GlcNAc(b1-6)]GalNAc(a1-</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [9] Man(a1-3)[Man(a1-6)]Man(a1-3)[Man(a1-6)]Man(a1-</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [10] GlcNAc6Ac(b1-4)Glc3Me(a1-</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ... (4990 more not shown)</span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # Unique structures: 5</span></span></code></pre></div>
<p>Notice that magical ‚Äú# Unique structures: 5‚Äù? That‚Äôs your performance booster right there!</p>
<p>Let‚Äôs verify this optimization is real:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Only 5 unique graphs are stored internally</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">attr</span>(large_struc, <span class="st">&quot;structures&quot;</span>))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 5</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># But we have 5,000 total elements</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(large_struc)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 5000</span></span></code></pre></div>
</div>
<div id="the-memory-savings-are-real" class="section level3">
<h3>The Memory Savings Are Real</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lobstr)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">obj_sizes</span>(struc, large_struc)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; * 14.33 kB</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; * 80.72 kB</span></span></code></pre></div>
<p><strong>80 kB vs 15 MB?</strong> That‚Äôs a 200x memory efficiency! But the real magic happens with computation speed‚Ä¶</p>
</div>
</div>
<div id="enter-the-smap-universe" class="section level2">
<h2>Enter the <code>smap</code> Universe üåå</h2>
<p>Now here‚Äôs the problem: if you try to use regular <code>lapply()</code> or <code>purrr::map()</code> functions on glycan structures, you‚Äôll hit a wall:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This won&#39;t work and will throw an error</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tryCatch</span>(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">map_int</span>(large_struc, <span class="sc">~</span> igraph<span class="sc">::</span><span class="fu">vcount</span>(.x)),</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">error =</span> <span class="cf">function</span>(e) <span class="fu">cat</span>(<span class="st">&quot;üí• Error:&quot;</span>, rlang<span class="sc">::</span><span class="fu">cnd_message</span>(e))</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; üí• Error: ‚Ñπ In index: 1.</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Caused by error in `ensure_igraph()`:</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ! Must provide a graph object (provided wrong object type).</span></span></code></pre></div>
<p><strong>Why does this fail?</strong> Because <code>purrr</code> functions don‚Äôt understand the internal structure optimization of <code>glycan_structure</code> objects.</p>
<div id="the-smap-family-to-the-rescue" class="section level3">
<h3>The <code>smap</code> Family to the Rescue!</h3>
<p>The <code>smap</code> functions (think ‚Äú<strong>s</strong>tructure map‚Äù) are drop-in replacements for <code>purrr</code> functions that are <strong>glycan-aware</strong>. They understand the unique structure optimization and work directly with the underlying graph objects.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This works beautifully!</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>vertex_counts <span class="ot">&lt;-</span> <span class="fu">smap_int</span>(large_struc, <span class="sc">~</span> igraph<span class="sc">::</span><span class="fu">vcount</span>(.x))</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>vertex_counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1] 5 2 3 5 2 5 2 3 5 2</span></span></code></pre></div>
<p><strong>The ‚Äús‚Äù stands for ‚Äústructure‚Äù</strong> ‚Äî these functions operate on the underlying <code>igraph</code> objects that represent your glycan structures.</p>
</div>
</div>
<div id="the-complete-smap-toolkit" class="section level2">
<h2>The Complete <code>smap</code> Toolkit üõ†Ô∏è</h2>
<p>The <code>smap</code> family provides glycan-aware equivalents for virtually all <code>purrr</code> functions:</p>
<table>
<thead>
<tr class="header">
<th>purrr</th>
<th>smap</th>
<th>purrr</th>
<th>smap</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>map()</code></td>
<td><code>smap()</code></td>
<td><code>map2()</code></td>
<td><code>smap2()</code></td>
</tr>
<tr class="even">
<td><code>map_lgl()</code></td>
<td><code>smap_lgl()</code></td>
<td><code>map2_lgl()</code></td>
<td><code>smap2_lgl()</code></td>
</tr>
<tr class="odd">
<td><code>map_int()</code></td>
<td><code>smap_int()</code></td>
<td><code>map2_int()</code></td>
<td><code>smap2_int()</code></td>
</tr>
<tr class="even">
<td><code>map_dbl()</code></td>
<td><code>smap_dbl()</code></td>
<td><code>map2_dbl()</code></td>
<td><code>smap2_dbl()</code></td>
</tr>
<tr class="odd">
<td><code>map_chr()</code></td>
<td><code>smap_chr()</code></td>
<td><code>map2_chr()</code></td>
<td><code>smap2_chr()</code></td>
</tr>
<tr class="even">
<td><code>some()</code></td>
<td><code>ssome()</code></td>
<td><code>pmap()</code></td>
<td><code>spmap()</code></td>
</tr>
<tr class="odd">
<td><code>every()</code></td>
<td><code>severy()</code></td>
<td><code>pmap_*()</code></td>
<td><code>spmap_*()</code></td>
</tr>
<tr class="even">
<td><code>none()</code></td>
<td><code>snone()</code></td>
<td><code>imap()</code></td>
<td><code>simap()</code></td>
</tr>
<tr class="odd">
<td></td>
<td></td>
<td><code>imap_*()</code></td>
<td><code>simap_*()</code></td>
</tr>
</tbody>
</table>
<p><strong>Simple rule</strong>: Replace <code>map</code> with <code>smap</code>, <code>pmap</code> with <code>spmap</code>, and <code>imap</code> with <code>simap</code>. Everything else works exactly like <code>purrr</code>!</p>
<div id="lets-put-them-to-work" class="section level3">
<h3>Let‚Äôs Put Them to Work!</h3>
<p><strong>Count vertices in each structure:</strong></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>vertex_counts <span class="ot">&lt;-</span> <span class="fu">smap_int</span>(large_struc, igraph<span class="sc">::</span>vcount)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(vertex_counts)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     2.0     2.0     3.0     3.4     5.0     5.0</span></span></code></pre></div>
<p><strong>Find structures with more than 4 vertices:</strong></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>has_many_vertices <span class="ot">&lt;-</span> <span class="fu">smap_lgl</span>(large_struc, <span class="sc">~</span> igraph<span class="sc">::</span><span class="fu">vcount</span>(.x) <span class="sc">&gt;</span> <span class="dv">4</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(has_many_vertices)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 2000</span></span></code></pre></div>
<p><strong>Get the degree sequence of each structure:</strong></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>degree_sequences <span class="ot">&lt;-</span> <span class="fu">smap</span>(large_struc, <span class="sc">~</span> igraph<span class="sc">::</span><span class="fu">degree</span>(.x))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>degree_sequences[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]  <span class="co"># Show first 3</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [[1]]</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 2 3 4 5 </span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 1 3 2 1 </span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [[2]]</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 2 </span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 1 </span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [[3]]</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 2 3 </span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 1 2</span></span></code></pre></div>
<p><strong>Check if any structure has isolated vertices:</strong></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ssome</span>(large_struc, <span class="sc">~</span> <span class="fu">any</span>(igraph<span class="sc">::</span><span class="fu">degree</span>(.x) <span class="sc">==</span> <span class="dv">0</span>))</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
<p><strong>Verify all structures are connected:</strong></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">severy</span>(large_struc, <span class="sc">~</span> igraph<span class="sc">::</span><span class="fu">is_connected</span>(.x))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div id="beyond-basic-smap" class="section level3">
<h3>Beyond Basic <code>smap()</code></h3>
<p>Quick examples of the extended family:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># smap2: Apply function with additional parameters</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>thresholds <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>large_enough <span class="ot">&lt;-</span> <span class="fu">smap2_lgl</span>(struc[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], thresholds, <span class="cf">function</span>(g, threshold) {</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  igraph<span class="sc">::</span><span class="fu">vcount</span>(g) <span class="sc">&gt;=</span> threshold</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>large_enough</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1]  TRUE FALSE FALSE</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># simap: Include position information</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>indexed_report <span class="ot">&lt;-</span> <span class="fu">simap_chr</span>(large_struc[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="cf">function</span>(g, i) {</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(<span class="st">&quot;#&quot;</span>, i, <span class="st">&quot;: &quot;</span>, igraph<span class="sc">::</span><span class="fu">vcount</span>(g), <span class="st">&quot; vertices&quot;</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>indexed_report</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;#1: 5 vertices&quot; &quot;#2: 2 vertices&quot; &quot;#3: 3 vertices&quot;</span></span></code></pre></div>
<p>‚ö†Ô∏è <strong>Performance Warning</strong>: <code>simap</code> functions don‚Äôt benefit from the unique structure optimization! Since each element has a different index, the combination of <code>(structure, index)</code> is always unique, breaking the deduplication that makes other <code>smap</code> functions fast. Use <code>simap</code> only when you truly need position information.</p>
</div>
</div>
<div id="performance-the-magic-of-deduplication" class="section level2">
<h2>Performance: The Magic of Deduplication ‚ö°</h2>
<p>The beauty of <code>smap</code> functions lies in automatic deduplication:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a large dataset with high redundancy</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>huge_struc <span class="ot">&lt;-</span> <span class="fu">rep</span>(struc, <span class="dv">5000</span>)  <span class="co"># 25,000 structures, only 5 unique</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Dataset size:&quot;</span>, <span class="fu">length</span>(huge_struc), <span class="st">&quot;structures</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Dataset size: 25000 structures</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Unique structures:&quot;</span>, <span class="fu">length</span>(<span class="fu">attr</span>(huge_struc, <span class="st">&quot;structures&quot;</span>)), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Unique structures: 5</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Redundancy factor:&quot;</span>, <span class="fu">length</span>(huge_struc) <span class="sc">/</span> <span class="fu">length</span>(<span class="fu">attr</span>(huge_struc, <span class="st">&quot;structures&quot;</span>)), <span class="st">&quot;x</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Redundancy factor: 5000 x</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tictoc)</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Optimized approach: smap only processes 5 unique structures</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>(<span class="st">&quot;smap_int (optimized)&quot;</span>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>vertex_counts_optimized <span class="ot">&lt;-</span> <span class="fu">smap_int</span>(huge_struc, igraph<span class="sc">::</span>vcount)</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; smap_int (optimized): 0.001 sec elapsed</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Naive approach: extract all graphs and process each one</span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a><span class="fu">tic</span>(<span class="st">&quot;Naive approach (all graphs)&quot;</span>)</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>all_graphs <span class="ot">&lt;-</span> <span class="fu">get_structure_graphs</span>(huge_struc)  <span class="co"># Extracts all 25,000 graphs</span></span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>vertex_counts_naive <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map_int</span>(all_graphs, igraph<span class="sc">::</span>vcount)</span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a><span class="fu">toc</span>()</span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Naive approach (all graphs): 0.087 sec elapsed</span></span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Verify results are equivalent (though data types may differ)</span></span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(vertex_counts_optimized, vertex_counts_naive)</span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p><strong>The higher the redundancy, the bigger the performance gain!</strong> In real glycoproteomics datasets with repeated structures, this optimization can provide about 10x speedups.</p>
</div>
<div id="advanced-patterns-and-tips" class="section level2">
<h2>Advanced Patterns and Tips üí°</h2>
<div id="working-with-complex-functions" class="section level3">
<h3>Working with Complex Functions</h3>
<p>The function you pass to <code>smap</code> must accept an <code>igraph</code> object as its first argument. You can use purrr-style lambda notation:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate clustering coefficient for each structure</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>clustering_coeffs <span class="ot">&lt;-</span> <span class="fu">smap_dbl</span>(large_struc, <span class="sc">~</span> igraph<span class="sc">::</span><span class="fu">transitivity</span>(.x, <span class="at">type =</span> <span class="st">&quot;global&quot;</span>))</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(clustering_coeffs)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA&#39;s </span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;       0       0       0       0       0       0    2000</span></span></code></pre></div>
</div>
<div id="combining-multiple-metrics" class="section level3">
<h3>Combining Multiple Metrics</h3>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a comprehensive analysis</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>structure_analysis <span class="ot">&lt;-</span> <span class="fu">smap</span>(large_struc, <span class="cf">function</span>(g) {</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">vertices =</span> igraph<span class="sc">::</span><span class="fu">vcount</span>(g),</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">edges =</span> igraph<span class="sc">::</span><span class="fu">ecount</span>(g),</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">diameter =</span> <span class="fu">ifelse</span>(igraph<span class="sc">::</span><span class="fu">is_connected</span>(g), igraph<span class="sc">::</span><span class="fu">diameter</span>(g), <span class="cn">NA</span>),</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">clustering =</span> igraph<span class="sc">::</span><span class="fu">transitivity</span>(g, <span class="at">type =</span> <span class="st">&quot;global&quot;</span>)</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to a more usable format</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>analysis_df <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, <span class="fu">lapply</span>(structure_analysis, data.frame))</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(analysis_df)</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   vertices edges diameter clustering</span></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1        5     4        3          0</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2        2     1        1        NaN</span></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3        3     2        1          0</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4        5     4        2          0</span></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5        2     1        1        NaN</span></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6        5     4        3          0</span></span></code></pre></div>
</div>
<div id="memory-efficient-filtering" class="section level3">
<h3>Memory-Efficient Filtering</h3>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find only structures with exactly 5 vertices</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>has_five_vertices <span class="ot">&lt;-</span> <span class="fu">smap_lgl</span>(large_struc, <span class="sc">~</span> igraph<span class="sc">::</span><span class="fu">vcount</span>(.x) <span class="sc">==</span> <span class="dv">5</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>five_vertex_structures <span class="ot">&lt;-</span> large_struc[has_five_vertices]</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Found&quot;</span>, <span class="fu">sum</span>(has_five_vertices), <span class="st">&quot;structures with exactly 5 vertices</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Found 2000 structures with exactly 5 vertices</span></span></code></pre></div>
</div>
</div>
<div id="when-to-use-smap-functions" class="section level2">
<h2>When to Use <code>smap</code> Functions</h2>
<p><strong>Use <code>smap</code> functions when:</strong></p>
<ul>
<li>‚úÖ You need to apply <code>igraph</code>-based functions to glycan structures</li>
<li>‚úÖ You want maximum performance with datasets containing repeated structures</li>
<li>‚úÖ You‚Äôre building custom glycan analysis pipelines</li>
</ul>
<p><strong>Stick with regular R functions when:</strong></p>
<ul>
<li>‚ùå Working with compositions</li>
<li>‚ùå Operating on string representations</li>
</ul>
<p>‚ö†Ô∏è <strong>Special note on <code>simap</code></strong>:</p>
<p>While <code>simap</code> functions are convenient for position-aware operations, they <strong>don‚Äôt provide performance benefits</strong> over regular <code>imap</code> functions. The inclusion of index information breaks the unique structure optimization, making each <code>(structure, index)</code> pair unique even when structures are identical.</p>
</div>
<div id="real-world-example-custom-motif-detection" class="section level2">
<h2>Real-World Example: Custom Motif Detection</h2>
<p>Here‚Äôs how you might build a custom glycan analysis pipeline using <code>smap</code> functions:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Custom motif detector</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>detect_branching <span class="ot">&lt;-</span> <span class="cf">function</span>(g) {</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  degrees <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">degree</span>(g)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">any</span>(degrees <span class="sc">&gt;=</span> <span class="dv">3</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply to large dataset - blazingly fast due to unique structure optimization</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>has_branching <span class="ot">&lt;-</span> <span class="fu">smap_lgl</span>(large_struc, detect_branching)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Structures with branching:&quot;</span>, <span class="fu">sum</span>(has_branching), <span class="st">&quot;out of&quot;</span>, <span class="fu">length</span>(large_struc), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Structures with branching: 2000 out of 5000</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Use smap2 to check structures against complexity thresholds</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>complexity_thresholds <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">4</span>), <span class="dv">1000</span>)  <span class="co"># Thresholds for each structure</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>meets_threshold <span class="ot">&lt;-</span> <span class="fu">smap2_lgl</span>(large_struc, complexity_thresholds, <span class="cf">function</span>(g, threshold) {</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>  igraph<span class="sc">::</span><span class="fu">vcount</span>(g) <span class="sc">&gt;=</span> threshold</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Structures meeting complexity threshold:&quot;</span>, <span class="fu">sum</span>(meets_threshold), <span class="st">&quot;out of&quot;</span>, <span class="fu">length</span>(large_struc), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Structures meeting complexity threshold: 2000 out of 5000</span></span></code></pre></div>
</div>
<div id="final-thoughts-youre-now-a-power-user" class="section level2">
<h2>Final Thoughts: You‚Äôre Now a Power User! üéâ</h2>
<p>Congratulations! You now understand the core optimization that makes <code>glyrepr</code> blazingly fast and how to leverage it with the <code>smap</code> family of functions.</p>
<p><strong>Key takeaways:</strong> - üß† <strong>Unique structure optimization</strong> is the secret sauce behind <code>glyrepr</code>‚Äôs performance - üöÄ <strong><code>smap</code> functions</strong> are drop-in replacements for <code>purrr</code> that understand glycan structures - ‚ö° <strong>Performance gains</strong> are dramatic with large datasets containing repeated structures - üõ†Ô∏è <strong>Use <code>smap</code> for structures</strong>, regular R functions for everything else</p>
<p>You‚Äôre now equipped to build the next generation of glycomics analysis tools. Go forth and analyze! üåü</p>
</div>
<div id="session-information" class="section level2">
<h2>Session Information</h2>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Running under: macOS Tahoe 26.0.1</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] C/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; tzcode source: internal</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; other attached packages:</span></span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] lobstr_1.1.2  dplyr_1.1.4   tibble_3.3.0  tictoc_1.2.1  purrr_1.1.0  </span></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [6] glyrepr_0.7.5</span></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1] vctrs_0.6.5       cli_3.6.5         knitr_1.50        rlang_1.1.6      </span></span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [5] xfun_0.53         stringi_1.8.7     generics_0.1.4    jsonlite_2.0.0   </span></span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [9] glue_1.8.0        prettyunits_1.2.0 backports_1.5.0   htmltools_0.5.8.1</span></span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [13] rstackdeque_1.1.1 sass_0.4.10       rmarkdown_2.30    evaluate_1.0.5   </span></span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [17] jquerylib_0.1.4   fastmap_1.2.0     yaml_2.3.10       lifecycle_1.0.4  </span></span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [21] stringr_1.5.2     compiler_4.5.1    igraph_2.2.0      pkgconfig_2.0.3  </span></span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [25] digest_0.6.37     R6_2.6.1          utf8_1.2.6        tidyselect_1.2.1 </span></span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [29] pillar_1.11.1     magrittr_2.0.4    bslib_0.9.0       checkmate_2.3.3  </span></span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [33] tools_4.5.1       cachem_1.1.0</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
